src/huggingface_hub/repocard_data.py

```diff
@@ -0,0 +1,369 @@
+        self.__dict__.update(kwargs)
+
+        if self.eval_results:
+            if type(self.eval_results) == EvalResult:
+                self.eval_results = [self.eval_results]
+            if self.model_name is None:
+                raise ValueError("`eval_results` requires `model_name` to be set.")
+
+    def to_dict(self):
+        """Converts CardData to a dict. It also formats the internal eval_results to
+        be compatible with the model-index format.
+
+        Returns:
+            `dict`: CardData represented as a dictionary ready to be dumped to a YAML
+            block for inclusion in a README.md file.
+        """
+
+        data_dict = copy.deepcopy(self.__dict__)
+        if self.eval_results is not None:
+            data_dict["model-index"] = eval_results_to_model_index(
+                self.model_name, self.eval_results
+            )
+            del data_dict["eval_results"], data_dict["model_name"]
+
+        return _remove_none(data_dict)
+
+    def to_yaml(self):
+        """Dumps CardData to a YAML block for inclusion in a README.md file."""
+        return yaml.dump(self.to_dict(), sort_keys=False).strip()
+
+    def __repr__(self):
+        return self.to_yaml()
+
+
+def model_index_to_eval_results(model_index: List[Dict[str, Any]]):
+    """Takes in a model index and returns a list of `huggingface_hub.EvalResult` objects.
+
+    A detailed spec of the model index can be found here:
+    https://github.com/huggingface/hub-docs/blob/main/modelcard.md
+
+    Args:
+        model_index (`List[Dict[str, Any]]`):
+        A model index data structure, likely coming from a README.md file on the
+        Hugging Face Hub.
+
+    Returns:
+        - model_name (`str`):
+            The name of the model as found in the model index. This is used as the
+            identifier for the model on leaderboards like PapersWithCode.
+        - eval_results (`List[EvalResult]`):
+            A list of `huggingface_hub.EvalResult` objects containing the metrics
+            reported in the provided model_index.
+
+    Example:
+        >>> from huggingface_hub.repocard_data import model_index_to_eval_results
+        >>> # Define a minimal model index
+        >>> model_index = [
+        ...     {
+        ...         "name": "my-cool-model",
+        ...         "results": [
+        ...         {
+        ...             "task": {
+        ...             "type": "image-classification"
+        ...             },
+        ...             "dataset": {
+        ...             "type": "beans",
+        ...             "name": "Beans"
+        ...             },
+        ...             "metrics": [
+        ...             {
+        ...                 "type": "accuracy",
+        ...                 "value": 0.9
+        ...             }
+        ...             ]
+        ...         }
+        ...         ]
+        ...     }
+        ... ]
+        >>> model_name, eval_results = model_index_to_eval_results(model_index)
```

Comment:
Do we really need to return the `model_name`? I would have expected this to return the `eval_results` only.

Reply:
Model name is required in the eval results.